sdks/go/node/core/containerruntime/docker/runContainer.go

```diff
@@ -82,9 +83,16 @@ func (cr _runContainer) RunContainer(
 	// do not change this prefix as it might break external consumers
 	containerName := getContainerName(req.ContainerID)
 	defer func() {
-		// ensure container always cleaned up
+		// ensure container always cleaned up: gracefully stop then delete it
+		newCtx := context.Background() // always use a fresh context, to clean up after cancellation
+		stopTimeout := 3 * time.Second
+		cr.dockerClient.ContainerStop(
```

Comment:
Maybe consider calling DeleteContainerIfExists here to consolidate the logic to a single place

Reply:
I gave that a shot, but found that we don't have the `_containerRuntime` in this context, just the `_runContainer`, which is a field of `_containerRuntime`. I don't really understand the split between these two structs: the docker client is present on both the parent and child structs and the parent struct has nothing else in it, so they seem redundant. The `RunContainer` method is present on both the `runContainer` interface and the `ContainerRuntime` interface and I don't understand how these relate to each other, if at all.

I'm not sure of the best way to refactor to consolidate the logic. My hunch is that the docker runtime needs to implement `RunContainer` directly on `_containerRuntime` and `_runContainer` needs to go away, but this sounds like a large refactor. Do you consider this blocking for this PR?

Reply:
They're just split for testability so we don't end up with a godmother struct. It's not blocking but just was an optimization. We can revisit at a later time. You might just mention in an @TODO comment they should be consolidated.

Reply:
Added the @TODO comment, please go ahead and merge when you have a chance. Thanks!